src/tools/rados/rados.cc

```diff
@@ -322,6 +328,94 @@ static int do_get(IoCtx& io_ctx, RadosStriper& striper,
   return ret;
 }
 
+static int do_repair_get(IoCtx& io_ctx, const char *objname,
+	const char *outfile, unsigned op_size, int32_t osdid, epoch_t epoch,
+	bool force)
+{
+  string oid(objname);
+
+  int fd;
+  if (strcmp(outfile, "-") == 0) {
+    fd = STDOUT_FILENO;
+  } else {
+    fd = TEMP_FAILURE_RETRY(::open(outfile, O_WRONLY|O_CREAT|O_TRUNC, 0644));
+    if (fd < 0) {
+      int err = errno;
+      cerr << "failed to open file: " << cpp_strerror(err) << std::endl;
+      return -err;
+    }
+  }
+
+  uint64_t offset = 0;
+  int ret = 0;
+  while (true) {
+    bufferlist outdata;
+    int flags = 0;
+    if (force) flags |= LIBRADOS_OP_FLAG_FAILOK;
+    ret = io_ctx.repair_read(oid, outdata, op_size, offset, flags, osdid, epoch);
+    if (ret <= 0) {
+      if (ret == -EINVAL || ret == -ERANGE)
+	cerr << "Specified epoch does not match scrub interval" << std::endl;
+      break;
+    }
+    ret = outdata.write_fd(fd);
+    if (ret < 0) {
+      cerr << "error writing to file: " << cpp_strerror(ret) << std::endl;
+      break;
+    }
+    if (outdata.length() < op_size)
+      break;
+    offset += outdata.length();
+  }
+
+  if (fd != STDOUT_FILENO)
+    VOID_TEMP_FAILURE_RETRY(::close(fd));
+  return ret;
+}
+
+static int do_repair_copy(IoCtx& io_ctx, const string& oid, const string& bad_shards_str,
+			  const char* epoch_str, const char* version_str, const string& what_str)
+{
+  vector<string> strs;
+  boost::split(strs, bad_shards_str, boost::is_any_of(","));
+  vector<librados::osd_shard_t> bad_shards;
+
+  for (const auto& s : strs) {
+    try {
+      auto delim = s.find('/');
+      osd_shard_t bad_shard;
+      if (delim == s.npos) {
+	bad_shard.osd = stoi(s);
+	bad_shard.shard = -1;
+      } else {
+	bad_shard.osd = stoi(s.substr(0, delim));
+	bad_shard.shard = stoi(s.substr(delim + 1));
+      }
+      bad_shards.push_back(bad_shard);
+    } catch (const exception& e) {
+      cerr << "Unable to parse the shards specified: " << strs
+	   << ". " << e.what() << std::endl;
+      return -EINVAL;
+    }
+  }
+  epoch_t epoch = atoi(epoch_str);
+  uint64_t version = atoi(version_str);
+  uint32_t what = (librados::repair_copy_t::DATA |
```

Comment:
Shouldn't we include "omap" in the default or just "data" as the default?  I feel like the command line format is limiting because you can only specify 1 of the elements.